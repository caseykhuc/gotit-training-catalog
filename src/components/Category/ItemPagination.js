import React from 'react';
import { Pagination } from 'react-bootstrap';
import PropTypes from 'prop-types';
import _ from 'lodash';

// return which pages should be shown currently
export const pages = (totalPages, currentPage, neighbors = 1) => {
  const pagesShown = neighbors * 2 + 1;

  if (totalPages <= pagesShown) {
    return _.range(totalPages);
  }

  if (currentPage <= neighbors) {
    return _.range(pagesShown);
  }

  if (currentPage >= totalPages - neighbors) {
    return _.range(totalPages - pagesShown, totalPages);
  }

  return _.range(currentPage - neighbors, currentPage + neighbors + 1);
}

// currentPage starts with 0
const ItemPagination = ({ currentPage, totalPages, onPageClick }) => (
  <Pagination className="justify-content-center m-4">
    <Pagination.First
      onClick={(e) => onPageClick(e, 0)}
      disabled={currentPage <= 0}
    />
    <Pagination.Prev
      onClick={(e) => onPageClick(e, currentPage - 1)}
      disabled={currentPage <= 0}
    />
    {pages(totalPages, currentPage).map((page) => (
      <Pagination.Item
        onClick={(e) => onPageClick(e, page)}
        active={page === currentPage}
        key={page}
      >
        {page + 1}
      </Pagination.Item>
    ))}
    <Pagination.Next
      onClick={(e) => onPageClick(e, currentPage + 1)}
      disabled={currentPage + 1 >= totalPages}
    />
    <Pagination.Last
      onClick={(e) => onPageClick(e, totalPages - 1)}
      disabled={currentPage + 1 >= totalPages}
    />
  </Pagination>
)

ItemPagination.propTypes = {
  totalPages: PropTypes.number.isRequired,
  currentPage: PropTypes.number.isRequired,
  onPageClick: PropTypes.func.isRequired,
};

export default ItemPagination;
