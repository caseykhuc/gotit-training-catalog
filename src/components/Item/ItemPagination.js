import React from 'react';
import { Pagination } from 'react-bootstrap';
import PropTypes from 'prop-types';

const ItemPagination = ({ currentPage, totalPages, onPageClick }) => {
  console.log(currentPage, totalPages); return (
    <Pagination>
      <Pagination.First
        onClick={(e) => onPageClick(e, 0)}
        disabled={currentPage <= 0}
      />
      <Pagination.Prev
        onClick={(e) => onPageClick(e, currentPage - 1)}
        disabled={currentPage <= 0}
      />
      {[...Array(totalPages)].map((page, i) => (
        <Pagination.Item
          onClick={(e) => onPageClick(e, i)}
          active={i === currentPage}
          key={i}
        >
          {i + 1}
        </Pagination.Item>
      ))}
      <Pagination.Next
        onClick={(e) => onPageClick(e, currentPage + 1)}
        disabled={currentPage + 1 >= totalPages}
      />
      <Pagination.Last
        onClick={(e) => onPageClick(e, totalPages - 1)}
        disabled={currentPage + 1 >= totalPages}
      />
    </Pagination>
  )
}

ItemPagination.propTypes = {
  totalPages: PropTypes.number.isRequired,
  currentPage: PropTypes.number.isRequired,
  onPageClick: PropTypes.func.isRequired,
};

export default ItemPagination;
