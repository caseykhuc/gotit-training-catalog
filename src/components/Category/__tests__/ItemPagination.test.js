import React from 'react';
import { shallow } from 'enzyme';
import { Pagination } from 'react-bootstrap';
import ItemPagination from '../ItemPagination';

describe('component/Category/ItemPagination', () => {
  let props;
  let wrapper;
  let pagiFirst;
  let pagiPrev;
  let pagiItem;
  let pagiNext;
  let pagiLast;

  const update = () => {
    wrapper.update();
    pagiFirst = wrapper.find(Pagination.First);
    pagiPrev = wrapper.find(Pagination.Prev);
    pagiItem = wrapper.find(Pagination.Item);
    pagiNext = wrapper.find(Pagination.Next);
    pagiLast = wrapper.find(Pagination.Last);
  };
  const setup = () => {
    wrapper = shallow(<ItemPagination {...props} />);
    update();
  };
  beforeEach(() => {
    props = {
      totalPages: 10,
      currentPage: 0,
      onPageClick: jest.fn(),
    }
  });
  it('should render correctly', () => {
    setup();
    expect(wrapper).toMatchSnapshot();

    props.currentPage = 5;
    setup();
    expect(wrapper).toMatchSnapshot();
  });
  it('should not render too much pages', () => {
    for (let i = 0; i < props.totalPages; i += 1) {
      props.currentPage = i;
      setup();
      expect(props.totalPages - pagiItem.length + 1).toBeTruthy();
    }
  });
  it('should invoke onPageClick properly', () => {
    props.currentPage = 3;
    setup();

    pagiFirst.simulate('click', {});
    expect(props.onPageClick).toBeCalledWith({}, 0);
    pagiPrev.simulate('click', {});
    expect(props.onPageClick).toBeCalledWith({}, props.currentPage - 1);
    pagiNext.simulate('click', {});
    expect(props.onPageClick).toBeCalledWith({}, props.currentPage + 1);
    pagiLast.simulate('click', {});
    expect(props.onPageClick).toBeCalledWith({}, props.totalPages - 1);
  });
})
