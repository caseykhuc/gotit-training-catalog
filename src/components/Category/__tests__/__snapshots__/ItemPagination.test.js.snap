// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component/Category/ItemPagination should render correctly 1`] = `ShallowWrapper {}`;

exports[`component/Category/ItemPagination should render correctly 2`] = `ShallowWrapper {}`;
